dimension/dimension.go

```diff
@@ -95,67 +111,106 @@ func (s *Store) GetUnique(w http.ResponseWriter, r *http.Request) {
+
+	log.InfoCtx(ctx, "added dimension to instance resource", logData)
+}
+
+func (s *Store) add(ctx context.Context, w http.ResponseWriter, r *http.Request, instanceID string, logData log.Data) (int, error) {
+	// Get instance
+	instance, err := s.GetInstance(instanceID)
+	if err != nil {
+		log.ErrorCtx(ctx, errors.WithMessage(err, "addDimensions endpoint: failed to GET instance"), logData)
+		return 0, err
+	}
+
 	// Early return if instance state is invalid
 	if err = models.CheckState("instance", instance.State); err != nil {
-		log.ErrorC("current instance has an invalid state", err, log.Data{"state": instance.State})
-		handleErrorType(errs.ErrInternalServer, w)
-		return
+		logData["state"] = instance.State
+		log.ErrorCtx(ctx, errors.WithMessage(err, "current instance has an invalid state"), logData)
+		return 0, err
 	}
 
 	option, err := unmarshalDimensionCache(r.Body)
 	if err != nil {
-		log.Error(err, nil)
-		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
+		log.ErrorCtx(ctx, err, logData)
+		return 400, err
 	}
-	option.InstanceID = id
+
+	option.InstanceID = instanceID
 	if err := s.AddDimensionToInstance(option); err != nil {
-		log.Error(err, nil)
-		handleErrorType(err, w)
+		log.ErrorCtx(ctx, err, logData)
+		return 0, err
 	}
+
+	return 201, nil
 }
 
-//AddNodeID against a specific value for dimension
-func (s *Store) AddNodeID(w http.ResponseWriter, r *http.Request) {
+// AddNodeIDHandler against a specific value for dimension
+func (s *Store) AddNodeIDHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
 	vars := mux.Vars(r)
-	id := vars["id"]
+	instanceID := vars["id"]
 	dimensionName := vars["dimension"]
 	value := vars["value"]
 	nodeID := vars["node_id"]
+	logData := log.Data{"instance_id": instanceID, "dimension_name": dimensionName, "option": value, "node_id": nodeID}
+	auditParams := common.Params{"instance_id": instanceID, "dimension_name": dimensionName, "option": value, "node_id": nodeID}
+
+	dim := models.DimensionOption{Name: dimensionName, Option: value, NodeID: nodeID, InstanceID: instanceID}
+
+	if err := s.addNodeID(ctx, w, r, dim, logData); err != nil {
+		s.Auditor.Record(ctx, PutNodeIDAction, audit.Unsuccessful, auditParams)
+		handleDimensionErr(ctx, err, 0, w, logData)
+		return
+	}
+
+	s.Auditor.Record(ctx, PutNodeIDAction, audit.Successful, auditParams)
+
+	log.InfoCtx(ctx, "added node id to dimension of an instance resource", logData)
+}
 
+func (s *Store) addNodeID(ctx context.Context, w http.ResponseWriter, r *http.Request, dim models.DimensionOption, logData log.Data) error {
 	// Get instance
-	instance, err := s.GetInstance(id)
+	instance, err := s.GetInstance(dim.InstanceID)
 	if err != nil {
-		log.ErrorC("Failed to GET instance when attempting to update a dimension of that instance.", err, log.Data{"instance": id})
+		log.ErrorC("Failed to GET instance when attempting to update a dimension of that instance.", err, logData)
 		handleErrorType(err, w)
```

Comment:
This function should no longer call `handlerErrorType` to set the status code on the response. The handler function that calls this also attempts to set the status code on the response. If this function has already set it, then it cannot be set again. Consider removing the `w` and `r` parameters from this function and just returning errors to be handled by the handler function.

Reply:
I will delete this function, it has been superseded.